File: DecisionTree.R
library(dplyr)
library(caret)
library(rpart)
library(rpart.plot)


#Read in and clean data
animals <- read.csv("animals-training.csv") %>%
  mutate_at(vars(Body.Covering, Animal), ~as.factor(.))
str(animals)
colnames(animals)[1] <- "Legs"

# create a train and test split. target variable (label) Animal. 70/30 split
index <- createDataPartition(y = animals$Animal,
                             p = .7,
                             list = FALSE)

animalsTrain <- animals[index,] 
animalsTest <- animals[-index,]



# Fit a decision tree
# Set the formula with all variables.If you choose, you can manually type the formula.
dtree <- rpart(Animal ~ Body.Covering + Legs, data = animalsTrain, method = "class", #categorical target
               control = rpart.control(minbucket = 2, cp = 0.00001, maxdepth = 7), 
               parms = list(split = "gini"))

# Plot the tree.
rpart.plot(dtree)



#----Training Error-------
predTrain <- predict(dtree, type = "class") 

#confusion matrix and accuracy
table(predTrain, animalsTrain$Animal)
(3+5+3+2+2)/(nrow(animalsTrain))



#----Testing Error-------
predTest <- predict(dtree, type = "class", newdata = animalsTest) 

#confusion matrix and accuracy
table(predTest, animalsTest$Animal)
(2)/(nrow(animalsTest))

File: DecisionTree.py
import pandas as pd
import os
import numpy as np
from sklearn import tree
from matplotlib import pyplot as plt

animals = pd.read_csv('animals-training.csv', header=0)
animals.head()

#Convert categroical variables to numberical by creating dummy variables
colList = list(animals.columns)
objList = []
for col in colList:
    if animals.dtypes[col] == object:
        objList.append(col)
objList # list of categorical variables


# 1-Hot encoding using pandas
animalsCats = pd.get_dummies(animals[objList[:-1]], prefix_sep='_', drop_first=False)
animalsCats.head()

#Get list of Continuous variables
contList = [item for item in colList if item not in objList] # select all columns not in categorical column list

#Create new dataframe with all continuous variables
aniamlsCont = animals[contList]
aniamlsVars = pd.concat([aniamlsCont, animalsCats],axis=1)
featureVals = aniamlsVars

#List of features to be used in model
features = list(aniamlsVars.columns)
features


#-----Create model--------
target = animals.iloc[:, -1].values    #Pull out the target variable

# train Decision Tree classifier
dTree = tree.DecisionTreeClassifier(max_leaf_nodes=None, criterion='gini') #Could also use entropy method
dTree = dTree.fit(featureVals,target)

#Determine accuracy of the model
predictions = dTree.predict(featureVals)
accuracy = (predictions == target).sum() / len(predictions)
accuracy

#Visualize the decision tree
plt.figure(figsize=(10,10))
tree.plot_tree(dTree, feature_names=(features), class_names=(np.unique(target)), filled=True)
plt.show()

